### PR TITLE
[next] Timestamp - Reader can see relative time of when that comment was posted

### PR DESCRIPTION
**Pull request by okbel**
### RelativeTime - Reader can see relative time of when that comment was posted
- Adds RelativeTime component
- Adds Timestamp component
- Adds "createdAt" field to the comment schema/resolvers
- Translations
- Show meaningful css names in dev
- Turn on antialiasing for font rendering
- Tests and test support for MatchMedia uses

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -16,5 +16,6 @@
   },
   "tslint.autoFixOnSave": true,
   "tslint.jsEnable": true,
-  "typescript.tsdk": "node_modules/typescript/lib"
+  "typescript.tsdk": "node_modules/typescript/lib",
+  "git.ignoreLimitWarning": true
```
Comment (alinhle) [inline .vscode/settings.json]: Can you turn it to false and see if you still have the issue?
Just making sure we really need this, it seems to make sense to keep the default.

File: src/core/client/framework/lib/bootstrap/TalkContext.tsx [hunk 2]
```diff
@@ -18,7 +23,7 @@ const { Provider, Consumer } = React.createContext<TalkContext>({} as any);
  */
 export const TalkContextConsumer = Consumer;
 
-/**
+/**s
```
Comment (alinhle) [inline src/core/client/framework/lib/bootstrap/TalkContext.tsx]: Here is a lonely `s`

File: src/core/client/stream/components/Comment.spec.tsx [hunk 3]
```diff
@@ -8,6 +8,7 @@ it("renders username and body", () => {
       username: "Marvin",
     },
     body: "Woof",
+    createdAt: new Date().toISOString(),
```
Comment (alinhle) [inline src/core/client/stream/components/Comment.spec.tsx]: This won't work well with jest snapshot, adding a fixed time would be better.

File: src/core/client/stream/components/Comment.spec.tsx [hunk 4]
```diff
@@ -20,6 +21,7 @@ it("renders with gutterBottom", () => {
...
```
Comment (alinhle) [inline src/core/client/stream/components/Comment.spec.tsx]: See above

File: src/core/client/framework/lib/bootstrap/createContext.tsx [hunk 5]
```diff
@@ -16,6 +19,19 @@ interface CreateContextArguments {
   init?: ((context: TalkContext) => void | Promise<void>);
 }
```
Comment (alinhle) [inline src/core/client/framework/lib/bootstrap/createContext.tsx:21]: We should add a comment here.

File: src/core/client/ui/components/Timestamp/Timestamp.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  expect(tree).toMatchSnapshot();
+});
+
+it("renders future date", () => {
+  const props = {
+    date: new Date("December 17, 2108 03:24:00").toISOString(),
+  };
+  const tree = create(<Timestamp {...props} />).toJSON();
+  expect(tree).toMatchSnapshot();
+});
```
Comment (alinhle) [inline src/core/client/ui/components/Timestamp/Timestamp.spec.tsx]: As suggested before, a test to check if consuming from the context is working correctly, would be helpful.

File: src/core/client/ui/components/Timestamp/Timestamp.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import { withStyles } from "talk-ui/hocs";
+import { PropTypesOf } from "talk-ui/types";
+
+import * as styles from "./Timestamp.css";
+
+interface InnerProps {
+  date: string;
+  live?: boolean;
+  classes: typeof styles;
+  formatter?: Formatter;
```
Comment (alinhle) [inline src/core/client/ui/components/Timestamp/Timestamp.tsx]: This is never used, shouldn't it be passed to `TimeAgo`?

File: src/locales/de/framework.ftl [hunk 8]
```diff
@@ -3,3 +3,37 @@
 ### among different targets.
 
 framework-validation-required = Dies ist ein Pflichtpfeld.
+
+framework-timeago =
```
Comment (alinhle) [inline src/locales/de/framework.ftl]: These German Translations are in Spanish?

File: src/locales/en-US/framework.ftl [hunk 9]
```diff
@@ -3,3 +3,38 @@
 ### among different targets.
 
 framework-validation-required = This field is required.
+
+framework-timeago =
+  { $value ->
```
Comment (alinhle) [inline src/locales/en-US/framework.ftl]: The `Spanish` translations is using `en 1 hora` but the English translation `1 hour from now`, I think it's better when we stay consistent and also translate to `in 1 hour`.

File: src/core/client/ui/components/Timestamp/Timestamp.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import TimeAgo, { Formatter } from "react-timeago";
+
+import { UIContext } from "talk-ui/components";
+import { withStyles } from "talk-ui/hocs";
+import { PropTypesOf } from "talk-ui/types";
+
+import * as styles from "./Timestamp.css";
+
+interface InnerProps {
+  date: string;
```
Comment (alinhle) [inline src/core/client/ui/components/Timestamp/Timestamp.tsx]: We should add support for `className` here.

File: src/core/client/ui/components/Timestamp/Timestamp.css [hunk 11]
```diff
@@ -0,0 +1,4 @@
+.root {
+  composes: button from "talk-ui/shared/typography.css";
```
Comment (alinhle) [inline src/core/client/ui/components/Timestamp/Timestamp.css]: We need to add typography styling from Sam or if already available from the Stylesheet

File: src/core/client/stream/components/Comment.tsx [hunk 12]
```diff
@@ -24,6 +25,7 @@ const Comment: StatelessComponent<CommentProps> = props => {
       <Typography className={styles.author} gutterBottom>
         {props.author && props.author.username}
       </Typography>
+      <Timestamp date={props.createdAt} />
```
Comment (alinhle) [inline src/core/client/stream/components/Comment.tsx]: We need to figure out a way to have `responsiveness` here.

**Review by alinhle — DISMISSED** [review]

**Comment by alinhle** [comment]
Also the large introspection files are still in the commit history, can you take another look?

File: src/core/client/stream/components/Comment.tsx (hunk 12 above)
Comment (alinhle) [reply src/core/client/stream/components/Comment.tsx]: I'm having a reusable component in mind, that we can always rely on.

File: src/locales/de/framework.ftl (hunk 8 above)
Comment (okbel) [reply src/locales/de/framework.ftl]: These are for you haha. I'm really rusty with german.

File: src/core/client/stream/components/Comment.tsx (hunk 12 above)
Comment (alinhle) [reply src/core/client/stream/components/Comment.tsx]: I'll work with you on this, I'm gonna test if one of my components work that I wrote in one of my earlier projects.

File: src/locales/en-US/framework.ftl (hunk 9 above)
Comment (okbel) [reply src/locales/en-US/framework.ftl]: 💅 : ok

Comment (alinhle) [reply src/locales/en-US/framework.ftl]: I tried that the last time actually and I first didn't know how to do it, but now I do, we can talk about it on Monday too :-)

File: src/core/client/ui/components/Timestamp/Timestamp.tsx (hunk 10 above)
Comment (okbel) [reply src/core/client/ui/components/Timestamp/Timestamp.tsx]: 👍

File: src/locales/en-US/framework.ftl (hunk 9 above)
Comment (alinhle) [reply src/locales/en-US/framework.ftl]: Actually in our current Talk we also changed from `1 hour from now` to `in 1 hour`, I don't know any reason who would use `1 hour from now`.

File: src/locales/de/framework.ftl (hunk 8 above)
Comment (alinhle) [reply src/locales/de/framework.ftl]: Haha sure

File: src/core/client/stream/components/Comment.tsx (hunk 12 above)
Comment (okbel) [reply src/core/client/stream/components/Comment.tsx]: 👍

File: src/core/client/ui/components/Timestamp/Timestamp.css (hunk 11 above)
Comment (okbel) [reply src/core/client/ui/components/Timestamp/Timestamp.css]: 👍

File: src/core/client/ui/components/Timestamp/Timestamp.tsx (hunk 7 above)
Comment (okbel) [reply src/core/client/ui/components/Timestamp/Timestamp.tsx]: This is old. Removed!

File: package.json [hunk 13]
```diff
@@ -45,6 +45,7 @@
     "mongodb": "^3.0.10",
     "passport": "^0.4.0",
     "performance-now": "^2.1.0",
+    "react-timeago": "^4.1.9",
```
Comment (wyattjoh) [inline package.json]: Should be in devDep instead of dep

File: src/core/client/framework/lib/bootstrap/TalkContext.tsx [hunk 14]
```diff
@@ -1,14 +1,19 @@
...
 
 export interface TalkContext {
   // relayEnvironment for our relay framework.
   relayEnvironment: Environment;
 
   // localMessages for our i18n framework.
   localeMessages: MessageContext[];
+
+  // formatter for timeago.
+  timeagoFormatter: Formatter | null;
```
Comment (wyattjoh) [inline src/core/client/framework/lib/bootstrap/TalkContext.tsx]: Why not use:

```typescript
timeagoFormatter?: Formatter;
```

instead?

**Review by wyattjoh — CHANGES_REQUESTED** [review]

File: package.json (hunk 13 above)
Comment (alinhle) [reply package.json]: 👍

File: src/core/client/framework/lib/bootstrap/TalkContext.tsx (hunk 14 above)
Comment (alinhle) [reply src/core/client/framework/lib/bootstrap/TalkContext.tsx]: I didn't want this to be optional, but to explicitly set to null in special cases e.g. for testing. Normally this is required.

**Review by wyattjoh — APPROVED** [review]

File: src/core/client/framework/lib/bootstrap/TalkContext.tsx (hunk 14 above)
Comment (alinhle) [reply src/core/client/framework/lib/bootstrap/TalkContext.tsx]: Thought about it again and made it optional :-P

**Pull request merged**